### DISCUSSION TITLE
Creating preview *and* production deployments from the `main` branch

### DISCUSSION THREAD

**Discussion by coding-velociraptor** [vercel/community#1601] [Help] [answered]
### Summary

I'm trying to deploy the `main` branch of my Github repository to Vercel in two ways: 1) On Push 2) On Release. I would like to deploy to a staging environment on every push, but deploy to production only on release. My current workaround is to have a mainly unused `release` branch, only pushed for a release. I have played with deploy hooks and a `vercel.json` configuration, however it is always bound to the branch. In my case both environments are on one branch. 

What do you suggest?

**Comment by amyegan** [comment, marked as answer]
The normal workflow for the Vercel for GitHub integration gives you preview deployments for PRs with production deployments on merge. So a separate `release` or `production` branch is a common strategy. 

But you can get more control over your workflow by [using GitHub Actions with Vercel CLI](https://vercel.com/guides/how-can-i-use-github-actions-with-vercel) to fully customize your release process with manual kickoff or [event-based triggers](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows)

**Reply by coding-velociraptor** [reply]
> So a separate release or production branch is a common strategy.

It's our current solution. However, it's not the preferred solution. Mainly, I don't think we need that extra branch. 

Is there any solution to use Vercel infra to build the assets? In the Vercel CLI approach, I would built on Github Actions.

**Reply by coding-velociraptor** [reply]
Another thing to consider: If I use the Vercel CLI on every push to `main`, will it associate the correct branch domain or only commit + author?

**Reply by amyegan** [reply]
CLI deployments still have branch association. Here's an example where I deployed from the `test` branch via CLI with local changes not committed to remote:
<img src="https://user-images.githubusercontent.com/8247105/221302821-3a220430-4e44-42d2-9356-561abde20231.png" alt="Project dashboard showing source as test branch without specific commit info deployed via CLI" />

And this is an example where I deployed from the `main` branch via CLI with the latest commit pulled from my remote branch:
<img src="https://user-images.githubusercontent.com/8247105/221586089-ab573108-a190-4cf3-8ca1-f08a15bd8d16.png" alt="Dashboard showing source as test branch with specific commit info deployed via CLI" />

Branch specific settings were used for each of these deployments and worked as expected 🙂

**Reply by coding-velociraptor** [reply]
Wonderful! Thank you for sharing the bit about the branch association. I will give it a shot and come back, if I encounter any problems with this approach.
